### ISSUE TITLE
Unable to npm install on windows

### ISSUE THREAD

**Issue by mavrick** [Voxer/sse4_crc32#29]
D:\>npm install sse4_crc32
\
> sse4_crc32@3.1.2 install D:\node_modules\sse4_crc32
> node-gyp rebuild

D:\node_modules\sse4_crc32>node "C:\Program Files\nodejs\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild
Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
  sse4_crc32.cpp
..\src\sse4_crc32.cpp(64): error C2065: '__asm__' : undeclared identifier [D:\node_modules\sse4_crc32\build\sse4_crc32.vcxproj]
..\src\sse4_crc32.cpp(64): error C2143: syntax error : missing ';' before 'volatile' [D:\node_modules\sse4_crc32\build\sse4_crc32.vcxproj]
..\src\sse4_crc32.cpp(64): error C2143: syntax error : missing ';' before 'volatile' [D:\node_modules\sse4_crc32\build\sse4_crc32.vcxproj]
..\src\sse4_crc32.cpp(189): warning C4244: '=' : conversion from 'unsigned __int64' to 'uint32_t', possible loss of data [D:\node_modules\sse4_crc32\build\sse4_crc32.vcxproj]
gyp ERR! build error
gyp ERR! stack Error: `C:\Windows\Microsoft.NET\Framework\v4.0.30319\msbuild.exe` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\build.js:267:23)
gyp ERR! stack     at ChildProcess.emit (events.js:98:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:820:12)
gyp ERR! System Windows_NT 6.1.7601
gyp ERR! command "node" "C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd D:\node_modules\sse4_crc32
gyp ERR! node -v v0.10.38
gyp ERR! node-gyp -v v1.0.1
gyp ERR! not ok

npm ERR! sse4_crc32@3.1.2 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the sse4_crc32@3.1.2 install script.
npm ERR! This is most likely a problem with the sse4_crc32 package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls sse4_crc32
npm ERR! There is likely additional logging output above.
npm ERR! System Windows_NT 6.1.7601
npm ERR! command "C:\\Program Files\\nodejs\\\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install" "sse4_crc32"
npm ERR! cwd D:\
npm ERR! node -v v0.10.38
npm ERR! npm -v 1.4.28
npm ERR! code ELIFECYCLE
npm ERR! not ok code 0

D:\>ver

Microsoft Windows [Version 6.1.7601]

**Comment by thanhlcm90** [comment]
Same problem too:

D:\NodeJs\store_management\sources\server_api\node_modules\sse4_crc32>node "C:\Program Files\nodejs\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild
child_process: customFds option is deprecated, use stdio instead.
Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
  sse4_crc32.cpp
..\src\sse4_crc32.cpp(64): error C2065: '__asm__' : undeclared identifier [D:\NodeJs\store_manage ment\sources\server_api\node_modules\sse4_crc32\build\sse4_crc32.vcxproj]
..\src\sse4_crc32.cpp(64): error C2143: syntax error : missing ';' before 'volatile' [D:\NodeJs\s tore_management\sources\server_api\node_modules\sse4_crc32\build\sse4_crc32.vcxproj]
..\src\sse4_crc32.cpp(189): warning C4244: '=' : conversion from 'unsigned __int64' to 'uint32_t' , possible loss of data [D:\NodeJs\store_management\sources\server_api\node_modules\sse4_crc32\bu ild\sse4_crc32.vcxproj]
gyp ERR! build error
gyp ERR! stack Error: `C:\Program Files (x86)\MSBuild\12.0\bin\msbuild.exe` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\build.js:267:23)
gyp ERR! stack     at ChildProcess.emit (events.js:110:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:1067:12)
gyp ERR! System Windows_NT 6.3.9600
gyp ERR! command "node" "C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd D:\NodeJs\store_management\sources\server_api\node_modules\sse4_crc32
gyp ERR! node -v v0.12.1
gyp ERR! node-gyp -v v1.0.2
gyp ERR! not ok
npm ERR! Windows_NT 6.3.9600
npm ERR! argv "C:\\Program Files\\nodejs\\\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install" "sse4_crc32"
npm ERR! node v0.12.1
npm ERR! npm  v2.5.1
npm ERR! code ELIFECYCLE

npm ERR! sse4_crc32@3.1.2 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the sse4_crc32@3.1.2 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the sse4_crc32 package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls sse4_crc32
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     D:\NodeJs\store_management\sources\server_api\npm-debug.log

**Comment by MystK** [comment]
Any fix for this? I am getting the same issue.

**Comment by anandsuresh** [comment]
Sorry guys but I haven't had a chance to do any kind of testing on Windows environments. And given my current work-load, I am not certain if I'll be able to get to it anytime soon. Feel free to send me a PR though.

**Comment by Thessalie** [comment]
Same problem here too. No gcloud in windows environment ?

**Comment by anandsuresh** [comment]
Given the number of people this is starting to affect, I'm going to take a peek into this issue over the upcoming weekend.

**Comment by pmarrapese** [comment]
Same problem here on Win7 x64:

```
H:\src\whym\merchant>npm install
npm WARN optional dep failed, continuing pk-grunt-shipit@^0.4.7
|


> sse4_crc32@3.1.2 install H:\src\whym\merchant\node_modules\sand-gcloud\node_modules\gcloud\node_modules\sse4_crc32
> node-gyp rebuild


H:\src\whym\merchant\node_modules\sand-gcloud\node_modules\gcloud\node_modules\sse4_crc32>if not defined npm_config_node_gyp (node "C:\Program Files\iojs\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (rebuild)
Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
  sse4_crc32.cpp
..\src\sse4_crc32.cpp(64): error C2065: '__asm__' : undeclared identifier [H:\src\whym\merchant\node_modules\sand-gcloud\node_modules\gcloud\node_modules\sse4_crc32\build\sse4_crc32.vcxproj]
..\src\sse4_crc32.cpp(64): error C2143: syntax error : missing ';' before 'volatile' [H:\src\whym\merchant\node_modules\sand-gcloud\node_modules\gcloud\node_modules\sse4_crc32\build\sse4_crc32.vcxproj]
..\src\sse4_crc32.cpp(64): error C2143: syntax error : missing ';' before 'volatile' [H:\src\whym\merchant\node_modules\sand-gcloud\node_modules\gcloud\node_modules\sse4_crc32\build\sse4_crc32.vcxproj]
..\src\sse4_crc32.cpp(184): error C3861: '_mm_crc32_u8': identifier not found [H:\src\whym\merchant\node_modules\sand-gcloud\node_modules\gcloud\node_modules\sse4_crc32\build\sse4_crc32.vcxproj]
..\src\sse4_crc32.cpp(189): error C3861: '_mm_crc32_u64': identifier not found [H:\src\whym\merchant\node_modules\sand-gcloud\node_modules\gcloud\node_modules\sse4_crc32\build\sse4_crc32.vcxproj]
..\src\sse4_crc32.cpp(191): error C3861: '_mm_crc32_u32': identifier not found [H:\src\whym\merchant\node_modules\sand-gcloud\node_modules\gcloud\node_modules\sse4_crc32\build\sse4_crc32.vcxproj]
..\src\sse4_crc32.cpp(192): error C3861: '_mm_crc32_u16': identifier not found [H:\src\whym\merchant\node_modules\sand-gcloud\node_modules\gcloud\node_modules\sse4_crc32\build\sse4_crc32.vcxproj]
..\src\sse4_crc32.cpp(193): error C3861: '_mm_crc32_u8': identifier not found [H:\src\whym\merchant\node_modules\sand-gcloud\node_modules\gcloud\node_modules\sse4_crc32\build\sse4_crc32.vcxproj]
gyp ERR! build error
gyp ERR! stack Error: `msbuild` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (C:\Program Files\iojs\node_modules\npm\node_modules\node-gyp\lib\build.js:269:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:169:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:1009:12)
gyp ERR! System Windows_NT 6.1.7601
gyp ERR! command "C:\\Program Files\\iojs\\node.exe" "C:\\Program Files\\iojs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd H:\src\whym\merchant\node_modules\sand-gcloud\node_modules\gcloud\node_modules\sse4_crc32
gyp ERR! node -v v1.8.1
gyp ERR! node-gyp -v v1.0.3
gyp ERR! not ok
npm ERR! Windows_NT 6.1.7601
npm ERR! argv "C:\\Program Files\\iojs\\node.exe" "C:\\Program Files\\iojs\\node_modules\\npm\\bin\\npm-cli.js" "install"
npm ERR! node v1.8.1
npm ERR! npm  v2.8.3
npm ERR! code ELIFECYCLE

npm ERR! sse4_crc32@3.1.2 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the sse4_crc32@3.1.2 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the sse4_crc32 package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls sse4_crc32
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     H:\src\whym\merchant\npm-debug.log
```

**Comment by Thessalie** [comment]
Thank you Anan ! On Windows 8 there's this problem about node-gyp trying to find msbuild.exe and failing each time occuring with any compiling module. The code is there but I dont have a clue on how to fix it C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\build.js:267:23

**Comment by ivan** [comment]
VS doesn't have an \_\_asm__, or even an __asm when compiling in 64-bit mode.  I think you have to use `__cpuid` instead:

https://msdn.microsoft.com/en-us/library/hskdteyh.aspx

GCC also has a way to get cpuid data without assembly code:

http://stackoverflow.com/questions/14266772/how-do-i-call-cpuid-in-linux/14266932#14266932

**Comment by anandsuresh** [comment]
@ivan Thanks for the links. I had to go the assembly route because we needed this to work on SmartOS/Solaris (our primary platform) on which I couldn't find any cpuid related documentation.

**Comment by anandsuresh** [comment]
@Thessalie @ivan @mavrick @thanhlcm90 @MystK @pmarrapese Can you please test #32 branch to see if that installs correctly?

Also, try running `node-e "console.log('H/w CRC Supported: ' + require('./sse4_crc32').isHardwareCrcSupported())"` and check if it returns true on your machines (assuming that H/W CRC is supported on your processors).

Thanks

**Comment by ivan** [comment]
Yep, that branch builds fine and I get `true` from `isHardwareCrcSupported()`.

 `.calculate()` seems to return the right results.

Thanks for the fix!

**Comment by anandsuresh** [comment]
@ivan Awesome! Will merge this change into master and release a new version.

**Comment by anandsuresh** [comment]
Fixed in release v3.2.0.

**Comment by Thessalie** [comment]
Congrats & thanks !

Le sam. 13 juin 2015 23:51, Anand Suresh <notifications@github.com> a
écrit :
